rfcs/015 - Reworked Compiler Front End.md

```diff
@@ -0,0 +1,179 @@
+
+The new compiler front-end will behave quite similarly:
+1. Define the value `abc123` with type `T`. Add a constraint that type `T` must match `Maybe<U>`, and that type `U` is `int64`.
+2. Find all functions and values named `none`, and use them as the list of possibilities for the value `abc123`.
+3. Define the value `abc124` with type `S`. Add a constraint that type `S` must match `string`.
+4. Find all functions and values named `toString`, and use them as the list of possibilities for the value `abc124`.
+5. Going back to the results of step 2, select `none` as defined in `root.ln`, since it is the only function that accepts 0 arguments.
+6. Going back to the results of step 4, select `toString(Maybe<Stringifiable>): string` as defined in `root.ln`, since it is the only function that matches the constraints on `T`.
+7. Inline all of the selected functions, and repeat steps 2-6 as necessary.
+
+Let's look at another example, which the current front-end can't solve without significant backtracking:
+```ln
+from @std/app import print, start, exit
+
+on start {
+  let myMaybe = none();
+  myMaybe = none();
+  myMaybe = some(5);
+  emit exit myMaybe.getOrExit().toInt8();
+}
+```
+
+There's a lot more complexity here!
+The current front-end would have to backtrack across all declarations, assignments, and Microstatement "rerefs" in order to ensure that the type assigned to `myMaybe` is `Maybe<int64>` once the `some(5)` assignment is reached.
+However, with the new front-end, the compiler would initially assign `myMaybe` to type `Maybe<T>`, then add a constraint that `T` can be assigned by the integer literal `5`.
+Then, once the `toInt8` call is reached, the compiler would add another constraint that for whatever type `T` is, there must be a function called `toInt8` that matches the signature `(T) -> int8`.
+If there are no other constraints, the compiler is able to resolve `T` as an `int64` before outputting the corresponding `amm`.
+
+The new compiler front-end will also be able to determine function signatures without requiring type annotations on parameters or the return type.
+A simple example might look like:
+```ln
+fn add4(num) {
+  return add(num, 4);
+}
+```
+
+At the time of writing this RFC, this function fails to parse due to a missing type annotation on `num`, but it's intended to support gradual typing even for function parameters in Alan.
+The current front-end will require substantial work to support this, requiring one of:
+- An interface (internal or in `root.ln`) that ensures there's an `add` operation for the type of `num` (for example, `Addable`).
+- Type-checking can only happen once the function is inlined.
+
+However, the new type resolution system would simply add a constraint that there be a function named `add` that satisfies the signature `(T, 4 as T) -> T` (ignoring the fact that arithmetic operations in Alan actually use `Result` values).
```

Comment:
I don't see how the `4 as T` is determined by the compiler? Since there is an add with the signature: `fn add(Result<int64>, int64): Result<int64>` already in the current codebase being able to assume that `num` and `4` are the same type does not seem correct?

But I don't think that affects the possibility of this particular example working or not.

Reply:
Ah, you're right - I'll fix that